java-extras/src/main/java/org/triplea/java/cache/ExpiringAfterWriteCache.java

```diff
@@ -0,0 +1,90 @@
+package org.triplea.java.cache;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import org.triplea.java.timer.ScheduledTimer;
+import org.triplea.java.timer.Timers;
+
+/**
+ * Cache that expires values when a TTL (time to live) expires. TTL timer starts when the value is
+ * written and is renewed if the value is 'refreshed'. The cache will reliably invoke a
+ * 'removeListener' at least once when cache items are removed or expired.
+ *
+ * @param <IdT> Type that identifies the keys of the map.
+ * @param <ValueT> Type that is placed as a value in the map.
+ */
+public class ExpiringAfterWriteCache<IdT, ValueT> implements TtlCache<IdT, ValueT> {
+
+  private final Cache<IdT, ValueT> cache;
+  private final ScheduledTimer cleanupTimer;
+  private final Consumer<CacheEntry<IdT, ValueT>> removalListener;
+
+  @SuppressWarnings("unchecked")
+  public ExpiringAfterWriteCache(
+      final long duration,
+      final TimeUnit timeUnit,
+      final Consumer<CacheEntry<IdT, ValueT>> removalListener) {
+    cache =
+        Caffeine.newBuilder()
+            .expireAfterWrite(duration, timeUnit)
+            .removalListener(
+                (key, value, cause) -> {
+                  if (cause == RemovalCause.EXPIRED || cause == RemovalCause.EXPLICIT) {
+                    removalListener.accept(new CacheEntry<>((IdT) key, (ValueT) value));
+                  }
+                })
+            .build();
+
+    this.removalListener = removalListener;
+
+    cleanupTimer =
+        Timers.fixedRateTimer("cache-cleanup-" + Math.random())
+            .period(1000, TimeUnit.MILLISECONDS)
+            .task(cache::cleanUp)
+            .start();
+  }
+
+  public void close() {
+    cleanupTimer.cancel();
+  }
+
+  @Override
+  public boolean refresh(final IdT id) {
+    final Optional<ValueT> value = get(id);
+
+    if (value.isPresent()) {
+      put(id, value.get());
+      return true;
+    }
+
+    return false;
+  }
+
+  @Override
+  public Optional<ValueT> get(final IdT id) {
+    return Optional.ofNullable(cache.getIfPresent(id));
+  }
+
+  @Override
+  public void put(final IdT id, final ValueT value) {
+    cache.put(id, value);
+  }
+
+  @Override
+  public Optional<ValueT> invalidate(final IdT id) {
+    final Optional<ValueT> value = get(id);
+    cache.invalidate(id);
+    value.ifPresent(valueT -> removalListener.accept(new CacheEntry<>(id, valueT)));
```

Comment:
Maybe I'm missing something, but why is removalListener called explicitly here?
Shouldn't invalidate trigger the listener by itself already?

Reply:
yes. Even if not, you can use `asMap().remove(id)` to remove the entry and obtain the old value atomically.

Reply:
@DanVanAtta If it's only the tests, then we could make the test wait until `ForkJoinPool.commonPool()` processed all events (I had a look at the source code), or alternatively just pass a custom executor for the test cases which simply runs the specified `Runnable`

Reply:
Reasonable suggestions, though I think sticking with the explicit notification is okay:
- in our context we handle duplicate calls already (at least we should be!). Having a contract that allows for duplicate callbacks seems okay in general as well (and if not then there is room for other implementations to change that).
- the test is already a bit slow at 700ms. Doing a join could mean we wait for the scheduled cleanup job which could add a second
- a join in the test would turn the test from mostly if not entirely a black box test into a white box test. Tests that do not rely on internal implementations tend to be easier to understand and maintain

A custom executor would also fall under the category of test scaffolding, it's better minimized. At the end of the day we also get benefit from an immediate callback, which does make the UI seem more responsive as games are removed nearly immediately and not after a small delay.